pkg/helm/option.go

```diff
@@ -383,10 +389,94 @@ func WithMaxHistory(max int32) HistoryOption {
 	}
 }
 
-// NewContext creates a versioned context.
+// NewContext creates a versioned context with kubernetes client data.
 func NewContext() context.Context {
-	md := metadata.Pairs("x-helm-api-client", version.Version)
-	return metadata.NewContext(context.TODO(), md)
+	return metadata.NewContext(
+		context.TODO(),
+		metadata.Join(
+			metadata.New(
+				extractKubeConfig()),
+			metadata.New(map[string]string{
+				"x-helm-api-client": version.Version,
+			}),
+		),
+	)
+}
+
+func extractKubeConfig() map[string]string {
+	configData := make(map[string]string)
+	clientConfig := cmdutil.DefaultClientConfig(pflag.NewFlagSet("", pflag.ContinueOnError))
+	c, err := clientConfig.ClientConfig()
```

Comment:
Why not just clean `c` of fields we don't want, serialize it to JSON, and then send that instead of creating a whole bunch of different headers? That would simplify both the marshal and unmarshal operations for this data.

gRPC will even base64 encode/decode this for you if you suffix the header `-bin`. https://godoc.org/google.golang.org/grpc/metadata#New